<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Giovanny Covarrubias-Pazaran" />

<meta name="date" content="2022-09-08" />

<title>Spatial modeling using the sommer package</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Spatial modeling using the sommer package</h1>
<h4 class="author">Giovanny Covarrubias-Pazaran</h4>
<h4 class="date">2022-09-08</h4>



<p>The sommer package was developed to provide R users with a powerful and reliable multivariate mixed model solver for different genetic (in diploid and polyploid organisms) and non-genetic analyses. This package allows the user to estimate variance components in a mixed model with the advantages of specifying the variance-covariance structure of the random effects, specifying heterogeneous variances, and obtaining other parameters such as BLUPs, BLUEs, residuals, fitted values, variances for fixed and random effects, etc. The core algorithms of the package are coded in C++ using the Armadillo library to optimize dense matrix operations common in the derect-inversion algorithms.</p>
<p>This vignette is focused on showing the capabilities of sommer to fit spatial models using the two dimensional splines models.</p>
<p><strong>SECTION 1: Introduction </strong></p>
<ol style="list-style-type: decimal">
<li>Background in tensor products</li>
</ol>
<p><strong>SECTION 2: Spatial models</strong></p>
<ol style="list-style-type: decimal">
<li>Two dimensional splines (multiple spatial components)</li>
<li>Two dimensional splines (single spatial component)</li>
<li>Spatial models in multiple trials at once</li>
</ol>
<div id="section-1-introduction" class="section level2">
<h2>SECTION 1: Introduction</h2>
<div id="backgrounds-in-tensor-products" class="section level3">
<h3>Backgrounds in tensor products</h3>
<p>TBD</p>
</div>
</div>
<div id="section-2-spatial-models" class="section level2">
<h2>SECTION 2: Spatial models</h2>
<div id="two-dimensional-splines-multiple-spatial-components" class="section level3">
<h3>1) Two dimensional splines (multiple spatial components)</h3>
<p>In this example we show how to obtain the same results than using the SpATS package. This is achieved by using the <code>spl2Db</code> function which is a wrapper of the <code>tpsmmb</code> function.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(sommer)</span></code></pre></div>
<pre><code>## Loading required package: Matrix</code></pre>
<pre><code>## Loading required package: MASS</code></pre>
<pre><code>## Loading required package: lattice</code></pre>
<pre><code>## Loading required package: crayon</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">data</span>(DT_yatesoats)</span>
<span id="cb6-2"><a href="#cb6-2"></a>DT &lt;-<span class="st"> </span>DT_yatesoats</span>
<span id="cb6-3"><a href="#cb6-3"></a>DT<span class="op">$</span>row &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">as.character</span>(DT<span class="op">$</span>row))</span>
<span id="cb6-4"><a href="#cb6-4"></a>DT<span class="op">$</span>col &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">as.character</span>(DT<span class="op">$</span>col))</span>
<span id="cb6-5"><a href="#cb6-5"></a>DT<span class="op">$</span>R &lt;-<span class="st"> </span><span class="kw">as.factor</span>(DT<span class="op">$</span>row)</span>
<span id="cb6-6"><a href="#cb6-6"></a>DT<span class="op">$</span>C &lt;-<span class="st"> </span><span class="kw">as.factor</span>(DT<span class="op">$</span>col)</span>
<span id="cb6-7"><a href="#cb6-7"></a></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co"># SPATS MODEL</span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co"># m1.SpATS &lt;- SpATS(response = &quot;Y&quot;,</span></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="co">#                   spatial = ~ PSANOVA(col, row, nseg = c(14,21), degree = 3, pord = 2),</span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co">#                   genotype = &quot;V&quot;, fixed = ~ 1,</span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co">#                   random = ~ R + C, data = DT,</span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co">#                   control = list(tolerance = 1e-04))</span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co"># </span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="co"># summary(m1.SpATS, which = &quot;variances&quot;)</span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="co"># </span></span>
<span id="cb6-17"><a href="#cb6-17"></a><span class="co"># Spatial analysis of trials with splines </span></span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="co"># </span></span>
<span id="cb6-19"><a href="#cb6-19"></a><span class="co"># Response:                   Y         </span></span>
<span id="cb6-20"><a href="#cb6-20"></a><span class="co"># Genotypes (as fixed):       V         </span></span>
<span id="cb6-21"><a href="#cb6-21"></a><span class="co"># Spatial:                    ~PSANOVA(col, row, nseg = c(14, 21), degree = 3, pord = 2)</span></span>
<span id="cb6-22"><a href="#cb6-22"></a><span class="co"># Fixed:                      ~1        </span></span>
<span id="cb6-23"><a href="#cb6-23"></a><span class="co"># Random:                     ~R + C    </span></span>
<span id="cb6-24"><a href="#cb6-24"></a><span class="co"># </span></span>
<span id="cb6-25"><a href="#cb6-25"></a><span class="co"># </span></span>
<span id="cb6-26"><a href="#cb6-26"></a><span class="co"># Number of observations:        72</span></span>
<span id="cb6-27"><a href="#cb6-27"></a><span class="co"># Number of missing data:        0</span></span>
<span id="cb6-28"><a href="#cb6-28"></a><span class="co"># Effective dimension:           17.09</span></span>
<span id="cb6-29"><a href="#cb6-29"></a><span class="co"># Deviance:                      483.405</span></span>
<span id="cb6-30"><a href="#cb6-30"></a><span class="co"># </span></span>
<span id="cb6-31"><a href="#cb6-31"></a><span class="co"># Variance components:</span></span>
<span id="cb6-32"><a href="#cb6-32"></a><span class="co">#                   Variance            SD     log10(lambda)</span></span>
<span id="cb6-33"><a href="#cb6-33"></a><span class="co"># R                 1.277e+02     1.130e+01           0.49450</span></span>
<span id="cb6-34"><a href="#cb6-34"></a><span class="co"># C                 2.673e-05     5.170e-03           7.17366</span></span>
<span id="cb6-35"><a href="#cb6-35"></a><span class="co"># f(col)            4.018e-15     6.339e-08          16.99668</span></span>
<span id="cb6-36"><a href="#cb6-36"></a><span class="co"># f(row)            2.291e-10     1.514e-05          12.24059</span></span>
<span id="cb6-37"><a href="#cb6-37"></a><span class="co"># f(col):row        1.025e-04     1.012e-02           6.59013</span></span>
<span id="cb6-38"><a href="#cb6-38"></a><span class="co"># col:f(row)        8.789e+01     9.375e+00           0.65674</span></span>
<span id="cb6-39"><a href="#cb6-39"></a><span class="co"># f(col):f(row)     8.036e-04     2.835e-02           5.69565</span></span>
<span id="cb6-40"><a href="#cb6-40"></a><span class="co"># </span></span>
<span id="cb6-41"><a href="#cb6-41"></a><span class="co"># Residual          3.987e+02     1.997e+01 </span></span>
<span id="cb6-42"><a href="#cb6-42"></a></span>
<span id="cb6-43"><a href="#cb6-43"></a><span class="co"># SOMMER MODEL</span></span>
<span id="cb6-44"><a href="#cb6-44"></a>m1.sommer &lt;-<span class="st"> </span><span class="kw">mmer</span>(Y<span class="op">~</span><span class="dv">1</span><span class="op">+</span>V<span class="op">+</span><span class="kw">spl2Db</span>(col,row, <span class="dt">nsegments =</span> <span class="kw">c</span>(<span class="dv">14</span>,<span class="dv">21</span>), <span class="dt">degree =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>), <span class="dt">penaltyord =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">what =</span> <span class="st">&quot;base&quot;</span>), </span>
<span id="cb6-45"><a href="#cb6-45"></a>                  <span class="dt">random =</span> <span class="op">~</span>R<span class="op">+</span>C<span class="op">+</span><span class="kw">spl2Db</span>(col,row, <span class="dt">nsegments =</span> <span class="kw">c</span>(<span class="dv">14</span>,<span class="dv">21</span>), <span class="dt">degree =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>), <span class="dt">penaltyord =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">what=</span><span class="st">&quot;bits&quot;</span>),</span>
<span id="cb6-46"><a href="#cb6-46"></a>                  <span class="dt">data=</span>DT, <span class="dt">tolParConv =</span> <span class="fl">1e-6</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>## Warning: fixed-effect model matrix is rank deficient so dropping 8 columns / coefficients</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">summary</span>(m1.sommer)<span class="op">$</span>varcomp</span></code></pre></div>
<pre><code>##                VarComp VarCompSE     Zratio Constraint
## R.Y-Y       125.928235  89.77330  1.4027360   Positive
## C.Y-Y        -7.789528  24.29529 -0.3206189   Positive
## A:fC.Y-Y      0.000000  19.09624  0.0000000   Positive
## A:fR.Y-Y      0.000000  15.87659  0.0000000   Positive
## A:fC.R.Y-Y    0.000000  21.42763  0.0000000   Positive
## A:C.fR.Y-Y   82.177296  92.28630  0.8904604   Positive
## A:fC.fR.Y-Y   0.000000  25.46390  0.0000000   Positive
## units.Y-Y   405.900386  90.48195  4.4859820   Positive</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># get the fitted values for the spatial kernel and plot</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="co"># ff &lt;- fitted.mmer(m1.sommer)</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co"># DT$fit &lt;- as.matrix(Reduce(&quot;+&quot;,ff$Zu[-c(1:2)])) </span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co"># lattice::levelplot(fit~row*col,data=DT)</span></span></code></pre></div>
</div>
<div id="two-dimensional-splines-single-spatial-component" class="section level3">
<h3>2) Two dimensional splines (single spatial component)</h3>
<p>To reduce the computational burden of fitting multiple spatial kernels <code>sommer</code> provides a single spatial kernel method through the <code>spl2Da</code> function. This as will be shown, can produce similar results to the more flexible model. Use the one that fits better your needs.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="co"># SOMMER MODEL</span></span>
<span id="cb11-2"><a href="#cb11-2"></a>m2.sommer &lt;-<span class="st"> </span><span class="kw">mmer</span>(Y<span class="op">~</span><span class="dv">1</span><span class="op">+</span>V, </span>
<span id="cb11-3"><a href="#cb11-3"></a>                  <span class="dt">random =</span> <span class="op">~</span>R<span class="op">+</span>C<span class="op">+</span><span class="kw">spl2Da</span>(col,row, <span class="dt">nsegments =</span> <span class="kw">c</span>(<span class="dv">14</span>,<span class="dv">21</span>), <span class="dt">degree =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>), <span class="dt">penaltyord =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>)),</span>
<span id="cb11-4"><a href="#cb11-4"></a>                  <span class="dt">data=</span>DT, <span class="dt">tolParConv =</span> <span class="fl">1e-6</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="kw">summary</span>(m1.sommer)<span class="op">$</span>varcomp</span></code></pre></div>
<pre><code>##                VarComp VarCompSE     Zratio Constraint
## R.Y-Y       125.928235  89.77330  1.4027360   Positive
## C.Y-Y        -7.789528  24.29529 -0.3206189   Positive
## A:fC.Y-Y      0.000000  19.09624  0.0000000   Positive
## A:fR.Y-Y      0.000000  15.87659  0.0000000   Positive
## A:fC.R.Y-Y    0.000000  21.42763  0.0000000   Positive
## A:C.fR.Y-Y   82.177296  92.28630  0.8904604   Positive
## A:fC.fR.Y-Y   0.000000  25.46390  0.0000000   Positive
## units.Y-Y   405.900386  90.48195  4.4859820   Positive</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># get the fitted values for the spatial kernel and plot</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="co"># ff &lt;- fitted.mmer(m2.sommer)</span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="co"># DT$fit &lt;- as.matrix(Reduce(&quot;+&quot;,ff$Zu[-c(1:2)])) </span></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="co"># lattice::levelplot(fit~row*col,data=DT)</span></span></code></pre></div>
</div>
<div id="spatial-models-in-multiple-trials-at-once" class="section level3">
<h3>3) Spatial models in multiple trials at once</h3>
<p>Sometimes we want to fit heterogeneous variance components when e.g., have multiple trials or different locations. The spatial models can also be fitted that way using the <code>at.var</code> and <code>at.levels</code> arguments. The first argument expects a variable that will define the levels at which the variance components will be fitted. The second argument is a way for the user to specify the levels at which the spatial kernels should be fitted if the user doesn’t want to fit it for all levels (e.g., trials or fields).</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>DT2 &lt;-<span class="st"> </span><span class="kw">rbind</span>(DT,DT)</span>
<span id="cb14-2"><a href="#cb14-2"></a>DT2<span class="op">$</span>Y &lt;-<span class="st"> </span>DT2<span class="op">$</span>Y <span class="op">+</span><span class="st"> </span><span class="kw">rnorm</span>(<span class="kw">length</span>(DT2<span class="op">$</span>Y))</span>
<span id="cb14-3"><a href="#cb14-3"></a>DT2<span class="op">$</span>trial &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;A&quot;</span>,<span class="kw">nrow</span>(DT)),<span class="kw">rep</span>(<span class="st">&quot;B&quot;</span>,<span class="kw">nrow</span>(DT)))</span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="kw">head</span>(DT2)</span></code></pre></div>
<pre><code>##   row col         Y   N          V  B         MP R C trial
## 1   1   1  91.67144 0.2    Victory B2    Victory 1 1     A
## 2   2   1  62.61736   0    Victory B2    Victory 2 1     A
## 3   3   1 120.90211 0.4 Marvellous B2 Marvellous 3 1     A
## 4   4   1 144.70494 0.6 Marvellous B2 Marvellous 4 1     A
## 5   5   1 149.66639 0.6 GoldenRain B2 GoldenRain 5 1     A
## 6   6   1 108.56940 0.2 GoldenRain B2 GoldenRain 6 1     A</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="co"># SOMMER MODEL</span></span>
<span id="cb16-2"><a href="#cb16-2"></a>m3.sommer &lt;-<span class="st"> </span><span class="kw">mmer</span>(Y<span class="op">~</span><span class="dv">1</span><span class="op">+</span>V, </span>
<span id="cb16-3"><a href="#cb16-3"></a>                  <span class="dt">random =</span> <span class="op">~</span><span class="kw">vsr</span>(<span class="kw">dsr</span>(trial),R)<span class="op">+</span><span class="kw">vsr</span>(<span class="kw">dsr</span>(trial),C)<span class="op">+</span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="st">                    </span><span class="kw">spl2Da</span>(col,row, <span class="dt">nsegments =</span> <span class="kw">c</span>(<span class="dv">14</span>,<span class="dv">21</span>), <span class="dt">degree =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">3</span>), <span class="dt">penaltyord =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">at.var =</span> trial),</span>
<span id="cb16-5"><a href="#cb16-5"></a>                  <span class="dt">rcov =</span> <span class="op">~</span><span class="kw">vsr</span>(<span class="kw">dsr</span>(trial),units),</span>
<span id="cb16-6"><a href="#cb16-6"></a>                  <span class="dt">data=</span>DT2, <span class="dt">tolParConv =</span> <span class="fl">1e-6</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb16-7"><a href="#cb16-7"></a><span class="kw">summary</span>(m3.sommer)<span class="op">$</span>varcomp</span></code></pre></div>
<pre><code>##              VarComp VarCompSE    Zratio Constraint
## A:R.Y-Y     105.9756  82.63815 1.2824054   Positive
## B:R.Y-Y      99.6331  81.42095 1.2236789   Positive
## A:C.Y-Y     145.8660 139.84741 1.0430368   Positive
## B:C.Y-Y     146.6229 140.81317 1.0412588   Positive
## A:all.Y-Y   373.3466 895.60046 0.4168674   Positive
## B:all.Y-Y   377.2377 909.03684 0.4149861   Positive
## A:units.Y-Y 402.2073 208.19856 1.9318448   Positive
## B:units.Y-Y 410.3528 211.60357 1.9392525   Positive</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># get the fitted values for the spatial kernel and plot</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="co"># ff &lt;- fitted.mmer(m3.sommer)</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="co"># DT2$fit &lt;- as.matrix(Reduce(&quot;+&quot;,ff$Zu[-c(1:4)])) </span></span>
<span id="cb18-4"><a href="#cb18-4"></a><span class="co"># lattice::levelplot(fit~row*col|trial,data=DT2)</span></span></code></pre></div>
</div>
</div>
<div id="literature" class="section level2">
<h2>Literature</h2>
<p>Covarrubias-Pazaran G. 2016. Genome assisted prediction of quantitative traits using the R package sommer. PLoS ONE 11(6):1-15.</p>
<p>Covarrubias-Pazaran G. 2018. Software update: Moving the R package sommer to multivariate mixed models for genome-assisted prediction. doi: <a href="https://doi.org/10.1101/354639" class="uri">https://doi.org/10.1101/354639</a></p>
<p>Bernardo Rex. 2010. Breeding for quantitative traits in plants. Second edition. Stemma Press. 390 pp.</p>
<p>Gilmour et al. 1995. Average Information REML: An efficient algorithm for variance parameter estimation in linear mixed models. Biometrics 51(4):1440-1450.</p>
<p>Henderson C.R. 1975. Best Linear Unbiased Estimation and Prediction under a Selection Model. Biometrics vol. 31(2):423-447.</p>
<p>Kang et al. 2008. Efficient control of population structure in model organism association mapping. Genetics 178:1709-1723.</p>
<p>Lee, D.-J., Durban, M., and Eilers, P.H.C. (2013). Efficient two-dimensional smoothing with P-spline ANOVA mixed models and nested bases. Computational Statistics and Data Analysis, 61, 22 - 37.</p>
<p>Lee et al. 2015. MTG2: An efficient algorithm for multivariate linear mixed model analysis based on genomic information. Cold Spring Harbor. doi: <a href="http://dx.doi.org/10.1101/027201" class="uri">http://dx.doi.org/10.1101/027201</a>.</p>
<p>Maier et al. 2015. Joint analysis of psychiatric disorders increases accuracy of risk prediction for schizophrenia, bipolar disorder, and major depressive disorder. Am J Hum Genet; 96(2):283-294.</p>
<p>Rodriguez-Alvarez, Maria Xose, et al. Correcting for spatial heterogeneity in plant breeding experiments with P-splines. Spatial Statistics 23 (2018): 52-71.</p>
<p>Searle. 1993. Applying the EM algorithm to calculating ML and REML estimates of variance components. Paper invited for the 1993 American Statistical Association Meeting, San Francisco.</p>
<p>Yu et al. 2006. A unified mixed-model method for association mapping that accounts for multiple levels of relatedness. Genetics 38:203-208.</p>
<p>Tunnicliffe W. 1989. On the use of marginal likelihood in time series model estimation. JRSS 51(1):15-27.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
